File: app/src/test/java/info/nightscout/androidaps/plugins/general/automation/actions/ActionLoopEnableTest.java
package info.nightscout.androidaps.plugins.general.automation.actions;

import com.google.common.base.Optional;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import info.AAPSMocker;
import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.R;
import info.nightscout.androidaps.interfaces.PluginType;
import info.nightscout.androidaps.interfaces.PumpDescription;
import info.nightscout.androidaps.plugins.aps.loop.LoopPlugin;
import info.nightscout.androidaps.plugins.configBuilder.ConfigBuilderPlugin;
import info.nightscout.androidaps.plugins.pump.virtual.VirtualPumpPlugin;
import info.nightscout.androidaps.queue.Callback;
import info.nightscout.androidaps.utils.SP;

import static org.powermock.api.mockito.PowerMockito.mock;
import static org.powermock.api.mockito.PowerMockito.when;

@RunWith(PowerMockRunner.class)
@PrepareForTest({MainApp.class, ConfigBuilderPlugin.class, SP.class})
public class ActionLoopEnableTest {
    ActionLoopEnable actionLoopEnable = new ActionLoopEnable();

    @Test
    public void friendlyNameTest() {
        Assert.assertEquals(R.string.enableloop, actionLoopEnable.friendlyName());
    }

    @Test
    public void shortDescriptionTest() {
        Assert.assertEquals(MainApp.gs(R.string.enableloop), actionLoopEnable.shortDescription());
    }

    @Test
    public void iconTest() {
        Assert.assertEquals(Optional.of(R.drawable.ic_play_circle_outline_24dp), actionLoopEnable.icon());
    }

    @Test
    public void doActionTest() {
        LoopPlugin.getPlugin().setPluginEnabled(PluginType.LOOP, false);
        actionLoopEnable.doAction(new Callback() {
            @Override
            public void run() {
            }
        });
        Assert.assertEquals(true, LoopPlugin.getPlugin().isEnabled(PluginType.LOOP));
        // another call should keep it enabled
        actionLoopEnable.doAction(new Callback() {
            @Override
            public void run() {
            }
        });
        Assert.assertEquals(true, LoopPlugin.getPlugin().isEnabled(PluginType.LOOP));
    }

    @Before
    public void prepareTest() {
        AAPSMocker.mockMainApp();
        AAPSMocker.mockApplicationContext();
        AAPSMocker.mockSP();
        AAPSMocker.mockConfigBuilder();
        AAPSMocker.mockCommandQueue();

        VirtualPumpPlugin pump = mock(VirtualPumpPlugin.class);
        when(pump.specialEnableCondition()).thenReturn(true);
        PumpDescription pumpDescription = new PumpDescription();
        pumpDescription.isTempBasalCapable = true;
        when(pump.getPumpDescription()).thenReturn(pumpDescription);
        when(AAPSMocker.configBuilderPlugin.getActivePump()).thenReturn(pump);

    }
}

File: app/src/test/java/info/nightscout/androidaps/plugins/general/automation/actions/ActionLoopEnableTest.kt
package info.nightscout.androidaps.plugins.general.automation.actions

import info.nightscout.androidaps.R
import info.nightscout.androidaps.interfaces.PluginType
import info.nightscout.androidaps.interfaces.PumpDescription
import info.nightscout.androidaps.plugins.bus.RxBusWrapper
import info.nightscout.androidaps.plugins.pump.virtual.VirtualPumpPlugin
import info.nightscout.androidaps.queue.Callback
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito
import org.mockito.Mockito.`when`
import org.powermock.core.classloader.annotations.PrepareForTest
import org.powermock.modules.junit4.PowerMockRunner

@RunWith(PowerMockRunner::class)
class ActionLoopEnableTest : ActionsTestBase() {

    lateinit var sut: ActionLoopEnable

    @Before
    fun setup() {

        `when`(virtualPumpPlugin.specialEnableCondition()).thenReturn(true)
        val pumpDescription = PumpDescription().apply { isTempBasalCapable = true }
        `when`(virtualPumpPlugin.pumpDescription).thenReturn(pumpDescription)
        `when`(activePlugin.activePump).thenReturn(virtualPumpPlugin)
        `when`(resourceHelper.gs(R.string.enableloop)).thenReturn("Enable loop")
        `when`(resourceHelper.gs(R.string.alreadyenabled)).thenReturn("Already enabled")

        sut = ActionLoopEnable(injector)
    }

    @Test fun friendlyNameTest() {
        Assert.assertEquals(R.string.enableloop, sut.friendlyName())
    }

    @Test fun shortDescriptionTest() {
        Assert.assertEquals("Enable loop", sut.shortDescription())
    }

    @Test fun iconTest() {
        Assert.assertEquals(R.drawable.ic_play_circle_outline_24dp, sut.icon())
    }

    @Test fun doActionTest() {
        `when`(loopPlugin.isEnabled(PluginType.LOOP)).thenReturn(false)
        sut.doAction(object : Callback() {
            override fun run() {}
        })
        Mockito.verify(loopPlugin, Mockito.times(1)).setPluginEnabled(PluginType.LOOP, true)
        Mockito.verify(configBuilderPlugin, Mockito.times(1)).storeSettings("ActionLoopEnable")

        `when`(loopPlugin.isEnabled(PluginType.LOOP)).thenReturn(true)

        // another call should keep it disabled, no new invocation
        sut.doAction(object : Callback() {
            override fun run() {}
        })
        Mockito.verify(loopPlugin, Mockito.times(1)).setPluginEnabled(PluginType.LOOP, true)
        Mockito.verify(configBuilderPlugin, Mockito.times(1)).storeSettings("ActionLoopEnable")
    }
}